cpp/src/gandiva/parse_url_holder.cc

```diff
@@ -0,0 +1,63 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "gandiva/parse_url_holder.h"
+#include "gandiva/node.h"
+
+namespace gandiva {
+
+    static bool IsArrowStringLiteral(arrow::Type::type type) {
+        return type == arrow::Type::STRING || type == arrow::Type::BINARY;
+    }
+
+    Status ParseUrlHolder::Make(const FunctionNode &node, std::shared_ptr <ParseUrlHolder> *holder) {
+        ARROW_RETURN_IF(node.children().size() != 2 && node.children().size() != 3,
+                        Status::Invalid("'parse_url' function requires two or three parameters"));
+
+        auto extract_part = dynamic_cast<LiteralNode*>(node.children().at(1).get());
+        ARROW_RETURN_IF(
+                extract_part == nullptr,
+                Status::Invalid("'parse_url' function requires a literal as the second parameter"));
+
+        auto extract_part_type = extract_part->return_type()->id();
+        ARROW_RETURN_IF(
+                !IsArrowStringLiteral(extract_part_type),
+                Status::Invalid(
+                        "'parse_url' function requires a string literal as the second parameter"));
+
+        if (node.children().size() == 3) {
+            auto pattern = dynamic_cast<LiteralNode*>(node.children().at(2).get());
+            ARROW_RETURN_IF(
+                    pattern == nullptr,
+                    Status::Invalid("'parse_url' function requires a literal as the third parameter"));
```

Comment:
Can we support non-literal case? If literal is enough in the workload, please also add the logic for checking literal in scala code and let the projection fall back for non-literal cases in `buildCheck`.

Reply:
It is a good question.After my local test is passed, I will consider supporting non-literal cases.

Reply:
I have changed the code to support non-literal cases.